[NEW GHOST ANTAG] The Paradox Clone

Shadyyy66 opened the pull request:
## About The Pull Request

![image](https://user-images.githubusercontent.com/114319683/200192566-e596ab83-8180-4f4c-a799-6e68e435fd43.png)

[DESIGN DOC]
https://hackmd.io/@tGKknGe5Q7qtddEspif79g/BJ3w9QLri/edit

Bluespace technology has destabilized the very fabric of reality, even time itself!

Sometimes during a shift, it's possible for a PARADOX CLONE to spawn somewhere on the station (usually maintenance). This paradox clone has the exact appearance and DNA as a random crew member on the station, wearing the same clothes they were at the start of the shift, including their ID.

To maintain timeline coherency, the paradox clone must now hunt down and kill the original and take their place or die trying, if the original survives they have failed.

The only thing the paradox clone spawns with is a mechanical toolbox that helps them from being stuck in maints if their ID doesn't have maints access.
## Why It's Good For The Game

The inclusion of this antag encourages good crew coordination and adds a less destructive, more personalized antag to the roster. 

This antag rewards good detective work, as both the clone and the original are identical in every way, and will therefore require a detective to break out their detective skills to figure out who the clone is.

Also a possible tie-in with the Chrono Legionnaires in the future.
## Changelog
:cl:
add: Time and space has destabilized even further due to Nanotrasen bluespace research, now, on occasion, clones from different timelines can appear on the station and hunt you down to maintain timeline stability! Beware the Paradox Clone!
/:cl:

Shadyyy66 committed:
Added Paradox Clone antag

not fully functional yet, but getting there, I think.

Shadyyy66 committed:
t

there was a t in preferences_savefile

Shadyyy66 committed:
fixing fixing

Shadyyy66 committed:
guhhh fixing guuhhhhh

Shadyyy66 committed:
Successful cloning achieved!!

Shadyyy66 committed:
Finalized!

The antag is now completely functional, woohoo!

tralezab commented:
kay based send it

MrDoomBringer commented:
Based but the whole 'spawning in maint / give them a toolbox' thing is silly just let the player choose where to spawn a la blob overmind

Shadyyy66 committed:
Update code/modules/events/ghost_role/paradox_clone_event.dm

Co-authored-by: 1bw0kopy <xz2rbf23@protonmail.com>

Shadyyy66 commented on code/modules/events/ghost_role/paradox_clone_event.dm:
```diff
@@ -0,0 +1,86 @@
+	playsound(S, 'sound/weapons/zapbang.ogg', 50, TRUE)
+	new /obj/item/storage/toolbox/mechanical(S.loc) //so they dont get stuck in maints
+
+	priority_announce("A time-space anomaly has been detected on the station, be aware of possible discrepancies.", "General Alert")
```
would i not want the spawning of the antag to be announced?

Shadyyy66 committed:
Update code/modules/events/ghost_role/paradox_clone_event.dm

Co-authored-by: 1bw0kopy <xz2rbf23@protonmail.com>

Pumpkinoe commented:
this is a cool idea, but my big question is how this stands out from Changeling, who can have similar objectives to escape with identity (and more tools available to it)

and also, does it inherent the quirks/traits of the player like blindness/deafness?

Shadyyy66 commented:
> this is a cool idea, but my big question is how this stands out from Changeling, who can have similar objectives to escape with identity (and more tools available to it)

thanks for asking.

It stands out from changeling by simply being a less lethal, less adaptive threat compared to the changeling. Unlike the changeling, the paradox clone can't change their appearance on the fly, and will be committed to that identity for the rest of the shift. Not only that, but the paradox clone has no obvious way to differentiate itself between clone and original (like how changelings have flesh clothes that dont come off, dead give-away) meaning there needs to be some detective work done to find out who the clone is.

> and also, does it inherent the quirks/traits of the player like blindness/deafness?

I realize now I haven't actually tested this yet, but seeing as it copies the targets DNA, i imagine it should copy their traits too, thanks for bringing it to my attention

GuillaumePrata commented:
Changeling has so much power build in that they often ignore all subterfuge and just armblade their target in a corner while using someone else's face, and even if caught, they can easily get away with monkey/slug/turning a corner and becoming someone else/armblading that they rarely try to convince people they are the real version.

Paradox clone has no build in power, anything they manage to do for their goals will be based on how creative the player can be and if they get caught, on how well they can talk it out.

moocowswag commented:
I cant find any flavor text suggesting if the paradox clone should avoid killing people that arnt their target or not.

This antag might become a minor issue for policy, especially on the lrp servers.

Shadyyy66 commented:
> I cant find any flavor text suggesting if the paradox clone should make an effort avoid killing people that arnt their target or not.

Their only objective is to kill their target, going out of your way to kill anyone unrelated would be murderboning.

Iamgoofball commented:
@Shadyyy66 that rule only applies on Manuel and very loosely; this will just be used as a free respawn w/ murder perms on terry/sybil/basil if you don't state otherwise

Ryusenshu commented:
> I cant find any flavor text suggesting if the paradox clone should make an effort avoid killing people that arnt their target or not.
> 
> This antag might become an issue for policy, especially on the lrp servers and worst-case scenario I fear that there will be cases based on admin discretion/bias if it is not addressed either by code (flavor text) or policy (headmin ruling)
> 
> (So if you don`t address it here with your code, we can move this debate to a policy thread)

cant even see this being much fun to be honest
seem like a cheap way to RR people and involving alot of meta info regarding behavior and such
imagine your target is part of a felinid gang, there is no point in trying to fool them

SELFHELL commented:
Its honestly a pretty cool thing,

Shadyyy66 committed:
Updated mission explaination

updated the mission explanation text to discourage antag murderboning

Iamgoofball commented:
``imagine your target is part of a felinid gang, there is no point in trying to fool them``
just ahelp when they metagame you for the easiest fucking ban on the planet

Shadyyy66 commented:
> @Shadyyy66 that rule only applies on Manuel and very loosely; this will just be used as a free respawn w/ murder perms on terry/sybil/basil if you don't state otherwise
> 
> please add flavor text saying to only kill the target

took this in mind and updated the objective explanation to discourage murderboning

blackdav123 commented:
what happens if you do win against your original? are you just a weak gearless antag cursed to roam the station and be killed by anyone witnessing your evil acts? 

this is a waste of threat and an antag roll

Shadyyy66 commented:
> what happens if you do win against your original? are you just a weak gearless antag cursed to roam the station and be killed by anyone witnessing your evil acts?
> 
> this is a waste of threat and an antag roll

I could add additional objectives if the current antag isnt sufficiently threatening or interesting enough, but for now, once you've killed your target, you just take over their job and act like a non-antag crewmate, like a Blood Brother after they've finished all of their objectives.

msgerbs commented:
> `imagine your target is part of a felinid gang, there is no point in trying to fool them` just ahelp when they metagame you for the easiest fucking ban on the planet

Which would they get banned on for knowing how their metagang members behave?

moocowswag commented:
> > `imagine your target is part of a felinid gang, there is no point in trying to fool them` just ahelp when they metagame you for the easiest fucking ban on the planet
> 
> Which rule would they get banned on for knowing how their metagang members behave?

none, because very basic relationships and knowledge of other players are allowed to persist as long as none of that knowledge results in you giving someone special benefits or meta`ing their behavior when they are an antagonist.

People meta me all the time as the "circuit lizard" so I can see instances of people getting screwed over if they disguised as me and didnt know how to do circuits (but i can metabreak that by pretending to be a noob)

LemonInTheDark commented:
> none, because very basic relationships and knowledge of other players are allowed to persist as long as none of that knowledge results in you giving someone special benefits or meta`ing their behavior when they are an antagonist.
> 
> People meta me all the time as the "circuit lizard" so I can see instances of people getting screwed over if they disguised as me and didnt know how to do circuits (but i can metabreak that by pretending to be a noob once in awhile.)

***special benefits***
admin issue anyhow. if they've set things up such that disguising is impossible that's a them problem

> what happens if you do win against your original? are you just a weak gearless antag cursed to roam the station and be killed by anyone witnessing your evil acts?
> 
> this is a waste of threat and an antag roll

seems fine to me, I like the lack of special gear, it vibes. nothing wrong with minor threats, gives people something to chew on. few things should be worldending
you might have trouble with people just taking it to murderbone, we've historically had issues with that. I trust arm tho.

necromanceranne commented:
I think there should be a 5% chance to spawn with a fake moustache (if you're not a plasmaman)

dragomagol commented on code/modules/events/ghost_role/paradox_clone_event.dm:
```diff
@@ -0,0 +1,85 @@
+	var/datum/mind/player_mind = new /datum/mind(selected.key)
+	player_mind.active = TRUE
+
+	var/mob/living/carbon/human/S = new ((pick(possible_spawns)))
```
Same here

dragomagol commented on code/modules/events/ghost_role/paradox_clone_event.dm:
```diff
@@ -0,0 +1,85 @@
+		return NOT_ENOUGH_PLAYERS
+
+	var/list/possible_spawns = list()
+	for(var/turf/X in GLOB.xeno_spawn)
```
No single-letter variables please

dragomagol commented on code/modules/events/ghost_role/paradox_clone_event.dm:
```diff
@@ -0,0 +1,85 @@
+		messenger.invisible = TRUE //clone doesnt show up on PDA message list
+
+	message_admins("[ADMIN_LOOKUPFLW(S)] has been made into a Paradox Clone by an event.")
+	S.log_message("was spawned as a Paradox Clone by an event.", LOG_GAME)
```
```suggestion
	S.log_message("was spawned as a Paradox Clone of [key_name(human_cloned)] by an event.", LOG_GAME)
```

dragomagol reviewed (commented):
I'm a fan. I could see it being more annoying if it was super common, but the paranoia is brings to the table, and having no obvious "tells," is really interesting to me. 

If people are acting in bad faith, that's an admin issue, not a code issue.

Shadyyy66 committed:
re-formatting + logging update

no more single letter variables and better logging information for when the antag is spawned

Shadyyy66 commented:
> also this needs to be a light impact dynamic ruleset, like obsession

why? they only have one objective, giving them a dynamic ruleset wont matter anyway

timothymtorres commented:
![chrome_IaI9qc3QAe](https://user-images.githubusercontent.com/5195984/200215287-66aad17d-521a-4cc6-a172-b8821f285607.png)

Shadowflame909 commented:
good way for a observers to get back onto station, and will be a fun thing for changelings to disguise themselves as to generate more terror. Definitely should be merged at a low threat cost (like obsessed)

dragomagol commented:
> I feel like "Evil Twin" is a more fighting name for the antag.

I like paradox clone, feels more sci-fi

Iamgoofball commented:
@dragomagol the better name is Mirror Universe You, per star trek tradition
give them a goatee and everything

Shadyyy66 committed:
Update code/modules/events/ghost_role/paradox_clone_event.dm

Co-authored-by: MrMelbert <51863163+MrMelbert@users.noreply.github.com>

Shadyyy66 committed:
Update code/modules/antagonists/paradox_clone/paradox_clone.dm

Co-authored-by: MrMelbert <51863163+MrMelbert@users.noreply.github.com>

Shadyyy66 committed:
Removed cloned captain station charter deletion

Mothblocks commented:
when the new antag has no design doc

Mothblocks commented:
@san7890 @RaveRadbury @Spookuni making sure you 3 are in the loop on a potential new antagonist being added

Mothblocks commented:
@tgstation/head-admins making sure you 3 are in the loop on a potential new antagonist being added

Shadyyy66 commented:
> when the new antag has no design doc

i mean i have a .txt file with some info in it, if i knew you wanted a design doc then i wouldve made one

Mothblocks commented:
we want them for anything really large like this, and we want it explaining specifically why the feature is the way it is and what you hope to accomplish more than what specifically it is: https://hackmd.io/@tgstation/BkzmU9EyK

feel free to send a hackmd afterwards

Spookuni commented:
I like it conceptually, and more minor low-impact antags are always good to see.

Colovorat commented:
Does toolbox include multitool? hacking with only wirecutters is pain in the ass, especially if grid has shit tone of power
Also if i read it correctly, it can only occure once? i think antag is not that imactful, so it will be fine having it spawn more than once per round?

Shadyyy66 commented:
> Does toolbox include multitool? hacking with only wirecutters is pain in the ass, especially if grid has shit tone of power Also if i read it correctly, it can only occure once? i think antag is not that imactful, so it will be fine having it spawn more than once per round?

The mechanical toolbox is for cutting through walls, not doors. And I'll think about making it possible to happen more than once.

msgerbs commented:
> > Does toolbox include multitool? hacking with only wirecutters is pain in the ass, especially if grid has shit tone of power Also if i read it correctly, it can only occure once? i think antag is not that imactful, so it will be fine having it spawn more than once per round?
> 
> The mechanical toolbox is for cutting through walls, not doors. And I'll think about making it possible to happen more than once.

Without a welding mask?

Shadyyy66 commented:
> Without a welding mask?

Yes. You can just power through it, you'll most likely just be cutting through 1-2 walls anyway.

Shadyyy66 commented:
> we want them for anything really large like this, and we want it explaining specifically why the feature is the way it is and what you hope to accomplish more than what specifically it is: https://hackmd.io/@tgstation/BkzmU9EyK
> 
> feel free to send a hackmd afterwards

decided to go ahead and make the design doc, if possible i want you to understand what im going for here 
https://hackmd.io/@tGKknGe5Q7qtddEspif79g/BJ3w9QLri

bluetshirtguy commented:
Oh boy, someone silently ganking people and putting no effort in their "disguise" just like what we already have in lings all the while policy complaints abound for what they can and can't do.

Mothblocks commented:
dont care about policy, thats the headmins job

Mothblocks commented:
> > we want them for anything really large like this, and we want it explaining specifically why the feature is the way it is and what you hope to accomplish more than what specifically it is: https://hackmd.io/@tgstation/BkzmU9EyK
> > feel free to send a hackmd afterwards
> 
> decided to go ahead and make the design doc, if possible i want you to understand what im going for here https://hackmd.io/@tGKknGe5Q7qtddEspif79g/BJ3w9QLri

403 for me, make sure you publish

Mothblocks reviewed (changes requested):
You need to make this a dynamic ruleset, random round events are more or less a backup and do not run in the vast vast majority of rounds

M3hillus commented:
Will they spawn in as the roundstart version of the target, or will they have a disguise "snapshot" in cases where someone is wearing a 1 of uniform? Lawyer/detective style suits come to mind.

Jackraxxus commented:
In the design document it's stated that "If there are any worries about sec possibly executing an innocent person due to a very clever clone play..". Surely this should be encouraged? Executioner is my favourite Town of Salem role. People who get executed over their clone being an exceptionally good liar should be told "skill issue, simply convince sec better next time."
Unless sec was especially dented in which case maybe "admin issue, simply Ahelp it" but I can't really think of a scenario outside of big time metagrudging where that would happen.

The doc goes on to view further potential problems with "If there’s ever a problem with the fact that someone has to be removed from the round permanently for this antag to succeed...". Surely, this is also based? Sure, it sucks to sit out of a long-form game, but where's the paranoia if the risk didn't exist? People who whine in deadchat about being round-removed by this antagonist should be told "cope", "mald", and "rule 10". In that order.
And even then, midround ghost roles like this one exist to get these people back into the round.

I'm, personally, very excited for this antagonist. I think it has big potential to add some very interesting deception/investigation gameplay. I'd hate for it to be watered-down.

And the idea of the paradox clone doing further evil than just replacing the target is a good one I think. Even if it's some light thief(I miss them so much)-tier shenanigans.

Mothblocks commented:
if you have concerns about the doc it'd be best if you post them on the doc where you can highlight waht you're talking about in a way thats easier for maintainers to digest

Mothblocks commented:
wait i cant comment on the doc what are your permission settings

M3hillus commented:
Will there be (changeling memories) to assist in copying the targets speech patterns?

Livrah commented:
Make it full antag once he killed the target.

DAKKA-WAAAGH commented:
Would be nice if it recycled the threat back into the round after it completes its objectives, so there's more room for bigger threats.

Also, I'm a huge fan of stealth and roleplay antagonism. **Very cool**.

Mothblocks commented:
@Livrah the codebase does not control policy. talk to the head admins. no more policy talk

Mothblocks commented:
Also no please do not put back in threat, one-off threats are completely fine and anticipated by Dynamic

Livrah commented:
Make the clone have more objectives once he killed the target.

Shadyyy66 committed:
Merge branch 'paradox_clone' of https://github.com/Shadyyy66/tgstation into paradox_clone

Jackal-boop commented:
I honestly quite like this in its current state.  Good work man

Mothblocks commented:
Ok it was just a hackmd hiccup. Please remove stuff out of the massive indent because it makes it impossible to comment (and also just isn't what we do)

realthomas69 commented:
this is a very unique idea, however the clone should get some new objectives after killing its target, or else it may quite often just end up being a waste of threat like thieves. we'll likely see the same boring thing playing out (guy silently kills target in maints, replaces him, shift proceeds as normal) happening far too often. Also, considering its a midround, it should absolutely gain a copy of the targets _current_ gear instead of their shiftstart gear. otherwise it could end up being pretty easy to pick out who the fake is

Mothblocks commented:
if you just give htem more objectives you have just added another antagonist that just chain murders and loses everything unique about them

Shadowflame909 commented:
> if you just give htem more objectives you have just added another antagonist that just chain murders and loses everything unique about them

What if it was a bizzaro objective. After usurping the life of their original. Their goal should be to do their job as horribly as possible. So a replaced sec officer would let prisoners go, a janitor would make the place dirtier. A cargo tech would strive to lose money.

It'd be a perfect middle round between grief, being bad at the job, and maybe not being who they say they are.

Mothblocks commented:
id love that but i have a feeling some of the people suggesting it wouldnt

Iamgoofball commented:
> or else it may quite often just end up being a waste of threat like thieves

you do not understand what an antagonist is

Shadyyy66 committed:
trying to add a dynamic midround ruleset and shit uhhh

Troxonater commented:
I like this antag, sounds fun.  As a suggestion though, I don't think assistant should be a job allowed to be affected by Paradox Clone. First of all, trying to track down your average assistant (who probably has suit sensors off) would be annoying, and I don't think people pay attention to assistant's enough for it to be a challenge to replace them other than the initial confrontation when you're trying to kill them without being called out.

Fikou commented:
how does it work with the intern id system?

Pinkufeck commented:
Waste of threat that impacts one person

Mothblocks commented:
the fun about the idea of having them then sabotage their department would be in that then it could affect more people without turning it into what people seemingly want it to be as another boring kill everyone you see antagonist

Iamgoofball commented:
@Pinkufeck you do not understand what an antagonist is

realthomas69 commented:
> if you just give them more objectives you have just added another antagonist that just chain murders and loses everything unique about them

they need some incentive or reason to create some kind of conflict. If you don't want them to become generic murder antags then don't give them generic murder objectives
@Mothblocks

Iamgoofball commented:
also you gamers literally whined so fucking hard about thieves because they weren't supposed to just murder everyone, so why the fuck are you gamers whining because we added an antagonist that kills someone

realthomas69 commented:
> you do not understand what an antagonist is

the point of an antagonist is to create conflict 
the evil twin becoming a normal crew member is obviously not a good way to create conflict

realthomas69 commented:
>they're someone pretending to be someone they aren't, that inherently creates conflict

This is absolutely true, but chances are high that simply nobody will notice and nothing will happen. I think it would be more interesting if it had some sort of goal or objective: conflict is what drives interesting roleplay and stories and it would be a shame if it just ended up doing nothing at all

>if you do not see this as creating conflict you have forgotten this is a roleplaying game, please quit playing it like top down TTT and then whining when people add things that don't conform to your absurd horse blinders view of how things should be
also you gamers literally whined so fucking hard about thieves because they weren't supposed to just murder everyone, so why the fuck are you gamers whining because we added an antagonist that kills someone

all I said was that I think this antagonist should have more objectives and you were somehow able to extrapolate all of this hidden meaning from that statement and then start raging at me for something I never said. Be civil!

Mothblocks commented:
> If you don't want them to become generic murder antags then don't give them generic murder objectives

right. which is why any new objectives they get should not just be "do the same thing but to someone else" in which case it will just be a guy who kills people and all flavor is useless

Addust commented:
this could react terribly if a paradox clone of a space explorer spawns, as they are forced to sit twiddling their thumbs praying that the explorer doesn't come back with traderoute gear, see them, and immediately go 'oh, a paradox clone' and dump a retro lasgun into them. if someone's off station Z, the chance of a paradox clone of them should be dropped dramatically to prevent this, so that a paradox clone can spawn when they return (only to get either killed on sight or bribed out of their greentext)

Shadyyy66 committed:
Added dynamic ruleset

Shadyyy66 committed:
ok it actually works now

Shadyyy66 committed:
added quirk cloning

have fun being a paraplegic clone

Shadyyy66 commented:
this antag SHOULD BE entirely functional now, let's get it merged 👏 👏

Shadyyy66 commented:
> You need to make this a dynamic ruleset, random round events are more or less a backup and do not run in the vast vast majority of rounds

this is done, the dynamic ruleset is functional

Bawhoppen commented:
This is the best antag idea in years

RikuTheKiller commented:
I like the idea, but it just seems like a respawn antag in some ways.

Ya know how changelings don't have any reason to really convince people they're the real deal since they have so many abilities? Well, what if we just give the clone abilities after it kills the original? And perhaps some kind of incentive to hurry the fuck up and kill them other than just telling them to kill them. (For example, after a certain duration, they destabilize out of existence.)

The duration thing would prevent them from just hiding their face, changing their clothes and just becoming your average crewmember. (Literally a fucking respawn antag at that point.)

Being able to just come into the round as someone else seems a little bit boring especially with how much effort it would take to silently kill the original. (Plus they would likely recognize you from speech patterns, perhaps give the clone a log of their target's speech so they know how they speak, at the very least.)

JohnFulpWillard commented:
hey your ci

Shadyyy66 committed:
Merge remote-tracking branch 'upstream/master' into paradox_clone

Shadyyy66 committed:
fuck you willard

code now updated to work properly with the new PDA system

Shadyyy66 committed:
added antag screenshot

Mothblocks commented:
You have png in your filename twice for the screenshot tests @Shadyyy66

Shadyyy66 committed:
sigh

fixed typo

Twaticus commented:
sorry missed the screenshot fail

theta060 commented:
"shoot him hes the clone"

JohnFulpWillard commented on code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm:
```diff
@@ -843,3 +843,82 @@
+	weight = 4
+	cost = 3
+	repeatable = TRUE
+	var/list/possible_spawns = list()
```
Doc please

JohnFulpWillard commented on code/modules/antagonists/paradox_clone/paradox_clone.dm:
```diff
@@ -0,0 +1,110 @@
+	preview_outfit = /datum/outfit/paradox_clone
+	count_against_dynamic_roll_chance = TRUE
+	//our target
+	var/datum/mind/original
```
The antag doesn't own the mind, so maybe this should be a weakref?

JohnFulpWillard commented on code/modules/antagonists/paradox_clone/paradox_clone.dm:
```diff
@@ -0,0 +1,110 @@
+	if(!original)//admins didn't set one
+		original = find_original()
+	if(!original)//we didn't find one
+		qdel(src)
```
This should stack_trace, this shouldn't happen at all should it?

JohnFulpWillard commented on code/modules/antagonists/paradox_clone/paradox_clone.dm:
```diff
@@ -0,0 +1,110 @@
+		chosen_victim = pick(possible_targets)
+	return chosen_victim
+
+/datum/objective/assassinate/paradox_clone
```
This should have a name so it's not confused with regular assassinate (from objective.dm ``//Name for admin prompts``)

JohnFulpWillard commented on code/modules/client/preferences/middleware/antags.dm:
```diff
@@ -118,6 +118,7 @@
 	var/static/list/non_ruleset_antagonists = list(
 		ROLE_FUGITIVE = /datum/antagonist/fugitive,
 		ROLE_LONE_OPERATIVE = /datum/antagonist/nukeop/lone,
+		ROLE_PARADOX_CLONE = /datum/antagonist/paradox_clone,
```
But this isn't the case, there is a ruleset for it

JohnFulpWillard commented on code/modules/events/ghost_role/paradox_clone_event.dm:
```diff
@@ -0,0 +1,85 @@
+/datum/round_event_control/paradox_clone
```
I don't know if I like this being an event on top of being a ruleset already, it's a lot of copy paste code and doesn't show up on the roundend report for threats, nor does it cost any.

KittyNoodle commented:
make a smite to sick one of these on a person

Shadyyy66 committed:
Update code/modules/antagonists/paradox_clone/paradox_clone.dm

Co-authored-by: John Willard <53777086+JohnFulpWillard@users.noreply.github.com>

Shadyyy66 committed:
Update code/modules/antagonists/paradox_clone/paradox_clone.dm

Co-authored-by: John Willard <53777086+JohnFulpWillard@users.noreply.github.com>

Shadyyy66 committed:
willard requested changes

removed random event, fixed wording, added weakref and stack_trace

Shadyyy66 committed:
Merge branch 'paradox_clone' of https://github.com/Shadyyy66/tgstation into paradox_clone

Shadyyy66 committed:
Merge branch 'master' into paradox_clone

Shadyyy66 committed:
PDA code updated

fixed PDA related error. stop changing the PDA code god damnit

JohnFulpWillard commented on code/modules/antagonists/paradox_clone/paradox_clone.dm:
```diff
@@ -0,0 +1,116 @@
+	if(!original_ref)//admins didn't set one
+		original_ref = WEAKREF(find_original())
+	if(!original_ref)//we didn't find one
+		stack_trace("No target found, aborting.")
```
```suggestion
		stack_trace("[src] was unable to find a target.")
```

JohnFulpWillard commented on code/modules/antagonists/paradox_clone/paradox_clone.dm:
```diff
@@ -0,0 +1,116 @@
+
+/datum/antagonist/paradox_clone/Destroy()
+	original_ref = null
+	..()
```
```suggestion
	return ..()
```

JohnFulpWillard commented on code/modules/antagonists/paradox_clone/paradox_clone.dm:
```diff
@@ -0,0 +1,116 @@
+
+	return clone_icon
+
+/datum/outfit/paradox_clone
```
outfit should be at the very bottom of the file so it's not mixed up with the antag's code

JohnFulpWillard commented on code/modules/antagonists/paradox_clone/paradox_clone.dm:
```diff
@@ -0,0 +1,116 @@
+	preview_outfit = /datum/outfit/paradox_clone
+	count_against_dynamic_roll_chance = TRUE
+
+	var/datum/weakref/original_ref ///antags target
```
```suggestion
	///Weakref to the clone's original, the target.
	var/datum/weakref/original_ref
```

JohnFulpWillard commented on code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm:
```diff
@@ -843,3 +843,84 @@
+	var/datum/mind/owner_mind = cloned.original_ref.resolve()
+	var/mob/living/carbon/human/human_cloned = owner_mind.current
+
+	clone.fully_replace_character_name(null, human_cloned.dna.real_name)
+	clone.name = human_cloned.name
+	human_cloned.dna.transfer_identity(clone, transfer_SE=1)
+	clone.age = human_cloned.age
+	clone.underwear = human_cloned.underwear
+	clone.undershirt = human_cloned.undershirt
+	clone.socks = human_cloned.socks
+	for(var/datum/quirk/target_quirk as anything in human_cloned.quirks)
+		clone.add_quirk(target_quirk.type)
+	clone.updateappearance(mutcolor_update=1)
+	clone.update_body()
+	clone.domutcheck()
+
+	//cloning clothing/ID/bag
+	clone.mind.assigned_role = human_cloned.mind.assigned_role
+
+	if(isplasmaman(human_cloned))
+		clone.equipOutfit(human_cloned.mind.assigned_role.plasmaman_outfit)
+		clone.internal = clone.get_item_for_held_index(1)
+	clone.equipOutfit(human_cloned.mind.assigned_role.outfit)
+
+	var/obj/item/clothing/under/sensor_clothes = clone.w_uniform
+	var/obj/item/modular_computer/pda/messenger = locate() in clone
+	if(messenger)
```
Would all this be better if it was on the antagonist itself? You find the target there, you could also make it mimic said target after that as well?

JohnFulpWillard commented on code/modules/antagonists/paradox_clone/paradox_clone.dm:
```diff
@@ -0,0 +1,116 @@
+	name = "\improper Paradox Clone"
+	roundend_category = "Paradox Clone"
+	job_rank = ROLE_PARADOX_CLONE
+	prevent_roundtype_conversion = FALSE
```
This var does nothing, so it doesn't do anything

![image](https://user-images.githubusercontent.com/53777086/206041578-54366a1f-5ce3-4d13-9d4d-cac72860acb7.png)

Someone should really delete it, it looks like a remnant of pre-dynamic

JohnFulpWillard commented on code/modules/antagonists/paradox_clone/paradox_clone.dm:
```diff
@@ -0,0 +1,116 @@
+	if(target?.current)
+		explanation_text = "Murder and replace [target.name], the [!target_role_type ? target.assigned_role.title : target.special_role]. Remember, your mission is to blend in, do not kill anyone else unless you have to!"
+	else
+		message_admins("WARNING! [ADMIN_LOOKUPFLW(owner)] paradox clone objectives forged without an original!")
```
stack trace this instead? That way we can see it in runtime logs (where its supposed to be because this isnt supposed to happen)

JohnFulpWillard commented on code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm:
```diff
@@ -843,3 +843,84 @@
+	weight = 4
+	cost = 3
+	repeatable = TRUE
+	var/list/possible_spawns = list() ///places the antag can spawn
```
```suggestion
	///List of places the antag can spawn
	var/list/possible_spawns = list()
```

JohnFulpWillard commented on code/modules/antagonists/paradox_clone/paradox_clone.dm:
```diff
@@ -0,0 +1,116 @@
+	var/list/possible_targets = list() //filters out silicons and simplemobs
+	var/chosen_victim  //The cloned player
+
+	for(var/mob/player as anything in GLOB.player_list)//prevents cloning of non-crew
+		if(!player.client || !player.mind || isnewplayer(player) || player.stat == DEAD || isbrain(player))
```
```suggestion
	for(var/mob/living/carbon/human/player as anything in GLOB.player_list)//prevents cloning of non-crew
		if(!player.client || !player.mind || player.stat == DEAD)
```

This will automatically check for humans, removing the need to check for brain and newplayer, while also making sure they don't copy a bot/simplemob/xenomorph

Shadyyy66 committed:
Update code/modules/antagonists/paradox_clone/paradox_clone.dm

Co-authored-by: John Willard <53777086+JohnFulpWillard@users.noreply.github.com>

Shadyyy66 committed:
Update code/modules/antagonists/paradox_clone/paradox_clone.dm

Co-authored-by: John Willard <53777086+JohnFulpWillard@users.noreply.github.com>

Shadyyy66 committed:
Update code/modules/antagonists/paradox_clone/paradox_clone.dm

Co-authored-by: John Willard <53777086+JohnFulpWillard@users.noreply.github.com>

Shadyyy66 committed:
Update code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm

Co-authored-by: John Willard <53777086+JohnFulpWillard@users.noreply.github.com>

Shadyyy66 committed:
Update code/modules/antagonists/paradox_clone/paradox_clone.dm

Co-authored-by: John Willard <53777086+JohnFulpWillard@users.noreply.github.com>

Shadyyy66 committed:
Removed prevent_roundtype_conversion

was told this is supposedly an obsolete variable

Shadyyy66 committed:
Merge branch 'paradox_clone' of https://github.com/Shadyyy66/tgstation into paradox_clone

JohnFulpWillard commented on code/modules/antagonists/paradox_clone/paradox_clone.dm:
```diff
@@ -0,0 +1,116 @@
+	var/list/possible_targets = list() //filters out silicons and simplemobs
+	var/chosen_victim  //The cloned player
+
+	for(var/mob/living/carbon/human/player as anything in GLOB.player_list)//prevents cloning of non-crew
```
```suggestion
	for(var/mob/living/carbon/human/player in GLOB.player_list)//prevents cloning of non-crew
```

I forgot to put this in my suggestion (it will make it actually check player_list for human)

SethLafuente commented:
Two major questions 

1.Will this antag be able to pick other antags and if it can will it have an uplink or have king abilities or be a rev/cultist 

2.Will this only apply to humans only I read something Fulp said and maybe I'm misinterpreting code but making it so it's a human only antag would be a unnecessarily limited decision for the antag

celulamp commented:
> > they're someone pretending to be someone they aren't, that inherently creates conflict
> 
> This is absolutely true, but chances are high that simply nobody will notice and nothing will happen. I think it would be more interesting if it had some sort of goal or objective: conflict is what drives interesting roleplay and stories and it would be a shame if it just ended up doing nothing at all
> 

I doubt anyone wants this but it would be cool if once you kiled someone you could interact with the body to create a new paradox clone once (for every clone) and you could replace a whole department eventually. (Basically better than changeling cus changeling fucking sucks lol)

but that's too high threat for this which should be a random event instead of a midround imo

SethLafuente commented:
> > > they're someone pretending to be someone they aren't, that inherently creates conflict
> 
> > 
> 
> > This is absolutely true, but chances are high that simply nobody will notice and nothing will happen. I think it would be more interesting if it had some sort of goal or objective: conflict is what drives interesting roleplay and stories and it would be a shame if it just ended up doing nothing at all
> 
> > 
> 
> 
> 
> I doubt anyone wants this but it would be cool if once you kiled someone you could interact with the body to create a new paradox clone once (for every clone) and you could replace a whole department eventually. (Basically better than changeling cus changeling fucking sucks lol)
> 
> 
> 
> but that's too high threat for this which should be a random event instead of a midround imo
> 
> 
> 
> 
> 
> 

We could have both one being just a single paradox clone taking over the other one the other a major antag event where there is 3 paradox clones and they start doing as you said from there adding an extra reason for conflict to appear for a paradox clone because it might be the bigger paradox event and not just a single anomaly

Overuseddart commented:
The paradox clone should get a heretic heart or similar to solve the problem of trying to track down the ssd assistant in the dorms 4# locker

SethLafuente commented:
> The paradox clone should get a heretic heart or similar to solve the problem of trying to track down the ssd assistant in the dorms 4# locker

It shouldn't be an object as it would be an obvious immediate killing crew there is a reason we no longer give heretics an actual heart and codex to have

Shadyyy66 committed:
Update code/modules/antagonists/paradox_clone/paradox_clone.dm

Co-authored-by: John Willard <53777086+JohnFulpWillard@users.noreply.github.com>

JohnFulpWillard commented:
> 2.Will this only apply to humans only I read something Fulp said and maybe I'm misinterpreting code but making it so it's a human only antag would be a unnecessarily limited decision for the antag

How are you expecting the antag to impersonate a cleanbot

Also not every antagonist needs to have deep mechanics, just do you jfc

Shadowflame909 commented:
> Also not every antagonist needs to have deep mechanics, just do you jfc

There's no shame in borrowing from already existing mechanics. Fugitives use the underused augmentation system to their advantage

But ultimately here the issue does not lie in a lack of tools but a lack of pressure it puts on the round. A similar gold standard of side antags like obsessed makes sure to get an entire job department involved in the madness with having to kill a coworker and what not. Here there might be minimal impact after replacing your target.

We need Idiocracy

Shadyyy66 committed:
moved cloning related code into antag datum

plus some minor code adjustments

Shadyyy66 committed:
Merge branch 'paradox_clone' of https://github.com/Shadyyy66/tgstation into paradox_clone

Fikou commented:
hows the progress on this its still drafted

Shadyyy66 committed:
Merge branch 'master' into paradox_clone

optimumtact commented:
We will test merge this after xmas period so it's not gunked up by present core gaming and if it goes well I dont' see a reason not to add it.

san7890 commented:
Forum Thread to Contain Gameplay-Related Feedback Here: https://tgstation13.org/phpBB/viewtopic.php?f=10&t=33043

@Shadyyy66 Do you have a forums account? I can switch you to be the author of this thread if you so desire.

tralezab commented:
I'm straight blushing over here that you made it similar to the obsessed. Absolutely kino antag

blackdav123 commented:
It would be cool as hell if more than one clone of a person could spawn at a time so you have two or even three of the same person hunting down the original and eachother in a free for all to become the only one

D4C-420 commented:
Just gonna copy paste is from the feedback threat since san said its worth mentioning here. 

1 .Clone has just looks like a generic version of the other person's job with nothing to blend in if the original has changed outfit, any items people know about, ect. This is makes it VERY clear who is real (the one with more stuff usually) and who is fake (the one with a set of tools and whatever they thought to take outta maint).
Maybe fix the above by letting the clone have the clothing of the original at the very least maybe the items, implants, genes ect. 

2. Clone does not get heirloom of whoever they copy (might be a bug, but still goes back to point 1 so I will mention it here anyway as a constant negative mood let sucks). Also if the original has a different kinda bag (leather satchel) they will be clearly different than the clone (default job bag)(again back to point one).

Shadyyy66 commented:
> 

1. I don't think it's a flaw that you have to make an effort to actually blend in, like all antags do, especially since blending in is what I'm trying to accomplish with this antag. Getting caught and unable to clear yourself of suspicion is you failing to blend in.

2. Yeah those sound like bugs I'll try and fix em

Helg2 commented:
If its the clone i think it should spawn with same implants as target (ms, tracker and etc.), or at least with ms, because officer/command without mindshield? You're coming to chapel right to crematory

SomeoneYouProbablyKnow commented:
I feel like the clone would benefit from a free mind reading mutation to mimic their target's speech patterns (Obviously undetectable). Lings get essentially the same effect on succing for a reason

DaydreamIQ commented:
Big issue I ran into while playing the clone. You don't have the same memories as your target, so you won't have an account number. Meaning anyone can just ask you what your account id is and sus you out for not knowing it.

Nebulacrity commented:
> I feel like the clone would benefit from a free mind reading mutation to mimic their target's speech patterns (Obviously undetectable). Lings get essentially the same effect on succing for a reason

Seconded. However, I suggest this not being as mundane as with Lings.

Instead, how about all verbal communications of the target - which, I assume, get recorded server-side for, at the very least, administration purposes - get perused by an algorithm which isolates series of strings that are close together due to them being part of either a monologue or dialogue. Then, when a Paradox Clone goes to sleep without having done so recently, they experience one of the isolated [mono/dia]logues instead of a randomly generated dream. Each isolated [mono/dia]logue can be experienced only once, and the isolation process can be re-done automatically every, say, 5-10 minutes to add more recent ones into the mix if the target decides to speak during that time as well.

And if there are no valid [mono/dia]logues left to experience? Just smush together one of the other, non-isolated (and thus non-[mono/dia]logue) strings of the target, much like with those normal random dreams. If none of those are valid either, then it's a random dream.

Bawhoppen commented:
@Nebulacrity Great idea on the speech pattern recognition algorithm. I look forward to you coding and PRing it.

Nebulacrity commented:
> @Nebulacrity Great idea on the speech pattern recognition algorithm. I look forward to you coding and PRing it.

Sorry, too busy nowadays. Y'all on your own. ;;]

Though it doesn't seem that difficult in theory - under the assumption that each logged speech string also contains data on when during the round it was said, the isolated groups are just those with time signature differences that are lower than [INSERT NUMBER HERE]. I'd probably go with 30 or 60 seconds, myself.

MrrFiish commented:
Just played a round as one and it was fun but I DO think that when you spawn it should copy their clothes and look at LEAST. When your target has hair dye and a ton of custom clothes then its practically impossible to find all the exact colors and items that they used to look good. It should also copy their bank ID because anyone can just ask both people and it figures out the clone.

Nebulacrity commented:
> Just played a round as one and it was fun but I DO think that when you spawn it should copy their clothes and look at LEAST. When your target has hair dye and a ton of custom clothes then its practically impossible to find all the exact colors and items that they used to look good. It should also copy their bank ID because anyone can just ask both people and it figures out the clone.

Yeah, I'm pretty sure it also doesn't copy the gender of the target which made figuring out who's the impostor really easy during that one round I was paying attention.

dragomagol commented:
#72437

dragomagol commented:
![image](https://user-images.githubusercontent.com/66640614/211132521-448a1cba-c95c-45f3-b03f-8bccfc05d6f9.png)

Time-Green commented:
Be really cool if it let you observe your target for a bit, so you can see how they're talking and dressed (because my guy was running around like a fucking caveman)

NamelessFairy commented:
Found a bug, if your a clone of someone with hair gradient you spawn without the gradient.

Bawhoppen commented:
DO NOT let this die

msgerbs commented:
Shady no!

Bawhoppen commented:
You can't do this to us

JohnFulpWillard committed:
Merge branch 'master' into pr/71141

JohnFulpWillard committed:
Updates Paradox clone to work with duplicate_object()

Paradox clones now use the duplicate_object() helper, and as well they use the "copy DNA" proc instead of the "Transfer Identity". This was practically the same, except copy dna was a little broken so I had to fix it up.
duplicate_object now properly transfers over quirks
I documented some things that were left behind, hopefully made things clearer to read, and also moved special role handling to on_gain and on

Because of how duplicate_object is, I had to move the actual duplication of the mob (and by extent, actually finding the target and setting the objective), to the midround ruleset instead of on the antag datum. This hurts Admin ability to give the antag out willy-nilly, but I'm sure admins can get special treatment later if someone is willing.

JohnFulpWillard committed:
Merge branch 'master' into paradox_clone

jlsnow301 commented on code/datums/dna.dm:
```diff
@@ -121,8 +121,7 @@ GLOBAL_LIST_INIT(total_uf_len_by_block, populate_total_uf_len_by_block())
 	new_dna.features = features.Copy()
-	new_dna.species = new species.type
-	new_dna.species.species_traits = species.species_traits
+	new_dna.holder.set_species(species.type, icon_update=0)
```
```suggestion
	new_dna.holder.set_species(species.type, icon_update = 0)
```

jlsnow301 commented on code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm:
```diff
@@ -844,3 +844,71 @@
+		JOB_HEAD_OF_SECURITY,
+		JOB_SECURITY_OFFICER,
+	)
+	required_enemies = list(2,2,1,1,1,1,1,0,0,0)
```
```suggestion
	required_enemies = list(2, 2, 1, 1, 1, 1, 1, 0, 0, 0)
```

jlsnow301 commented on code/modules/antagonists/paradox_clone/paradox_clone.dm:
```diff
@@ -0,0 +1,97 @@
+
+/datum/objective/assassinate/paradox_clone/update_explanation_text()
+	. = ..()
+	if(!target || !target.current)
```
```suggestion
	if(!target?.current)
```

jlsnow301 reviewed (changes requested):
otherwise this looks pretty good, will get to it today. sorry about the nits

Shadyyy66 committed:
Update code/datums/dna.dm

Co-authored-by: Jeremiah <42397676+jlsnow301@users.noreply.github.com>

Shadyyy66 committed:
Update code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm

Co-authored-by: Jeremiah <42397676+jlsnow301@users.noreply.github.com>

Shadyyy66 committed:
Update code/modules/antagonists/paradox_clone/paradox_clone.dm

Co-authored-by: Jeremiah <42397676+jlsnow301@users.noreply.github.com>

jlsnow301 commented:
Great work

msgerbs commented:
The absolute chads

Protosaurus0 commented:
Server: Fulpstation round  24440
Revision: BYOND Version: 514.1581
Server revision compiled on: 2023-02-05T09:29:55+00:00
Master commit: [94ed04224c27bb264c3367ec440fe9a1a6285b48](https://github.com/fulpstation/fulpstation/commit/94ed04224c27bb264c3367ec440fe9a1a6285b48)
Problem Description: Playing as paradox clone and I could attack very quickly(Like fist of the north star) & couldn't be slipped/stunned(like a hulk), I got these abilities after I killed my target/destroyed their brain and they got revived by in a pod-person plant. I later died and got revived, when taken off stasis my body was stuck with the stasis status & I couldn't use items/move despite being healthy enough to do so.
What did you expect to happen: Normal attack speed/stamina when playing paradox clone
What happened instead: Had stun resist/slip immunity + super fast attacks as paradox clone.
Why is this bad/What are the consequences: Could be abused, makes paradox clone nearly invincible if they're not trying to die.
Steps to reproduce the problem: Become paradox clone, kill target, destroy brain of target, revive target with a pod from botany. I'm unsure what the trigger was 


Sorry if this is in the wrong format/place.

LemonInTheDark commented:
Go here, it'll get put into the issue tracker
https://github.com/tgstation/tgstation/issues/new?assignees=&labels=&template=bug_report.md
